google/cloud/bigtable/internal/convert_policies_test.cc

```diff
@@ -0,0 +1,190 @@
+      options.has<bigtable_admin::BigtableInstanceAdminBackoffPolicyOption>());
+  EXPECT_TRUE(
+      options.has<bigtable_admin::BigtableInstanceAdminPollingPolicyOption>());
+}
+
+TEST(ConvertPolicies, TableAdmin) {
+  auto r = DefaultRPCRetryPolicy(internal::kBigtableTableAdminLimits);
+  auto b = DefaultRPCBackoffPolicy(internal::kBigtableTableAdminLimits);
+  auto p = DefaultPollingPolicy(internal::kBigtableTableAdminLimits);
+  auto options = bigtable_internal::MakeTableAdminOptions(
+      std::move(r), std::move(b), std::move(p));
+  EXPECT_TRUE(options.has<google::cloud::internal::GrpcSetupOption>());
+  EXPECT_TRUE(options.has<google::cloud::internal::GrpcSetupPollOption>());
+  EXPECT_TRUE(
+      options.has<bigtable_admin::BigtableTableAdminRetryPolicyOption>());
+  EXPECT_TRUE(
+      options.has<bigtable_admin::BigtableTableAdminBackoffPolicyOption>());
+  EXPECT_TRUE(
+      options.has<bigtable_admin::BigtableTableAdminPollingPolicyOption>());
+}
+
+/**
+ * This test uses custom policies to ensure that:
+ *
+ * 1. The retry and backokff `Setup()` function is called by invoking the
+ *    `GrpcSetupOption` and not the `GrpcSetupPollOption`.
+ *
+ * 2. The polling `Setup()` function is called by invoking the
+ *    `GrpcSetupPollOption` and not the `GrpcSetupOption`.
+ *
+ * 3. `GrpcSetupOption` and `GrpcSetupPollOption` use the initial state of the
+ *    policies to setup the `grpc::ClientContext`.
+ */
+TEST(ConvertPolicies, GrpcSetup) {
+  class RetryWithSetup : public RPCRetryPolicy {
+   public:
+    explicit RetryWithSetup(bool* called) : called_(called) {
+      *called_ = false;
+    }
+    std::unique_ptr<RPCRetryPolicy> clone() const override {
+      return absl::make_unique<RetryWithSetup>(called_);
+    }
+    void Setup(grpc::ClientContext&) const override {
+      EXPECT_FALSE(*called_)
+          << "GrpcSetupOption is not using the retry policy's initial state to "
+             "configure the context";
+      *called_ = true;
+    }
+    bool OnFailure(Status const&) override { return false; }
+    bool OnFailure(grpc::Status const&) override { return false; }
+
+   private:
+    bool* called_;
+  };
+
+  class BackoffWithSetup : public RPCBackoffPolicy {
+   public:
+    explicit BackoffWithSetup(bool* called) : called_(called) {
+      *called_ = false;
+    }
+    std::unique_ptr<RPCBackoffPolicy> clone() const override {
+      return absl::make_unique<BackoffWithSetup>(called_);
+    }
+    void Setup(grpc::ClientContext&) const override {
+      EXPECT_FALSE(*called_)
+          << "GrpcSetupOption is not using the backoff policy's initial state "
+             "to configure the context";
+      *called_ = true;
+    }
+    std::chrono::milliseconds OnCompletion(Status const&) override {
+      return {};
+    };
+    std::chrono::milliseconds OnCompletion(grpc::Status const&) override {
+      return {};
+    }
+
+   private:
+    bool* called_;
+  };
+
+  class PollingWithSetup : public PollingPolicy {
```

Comment:
This is what mock classes were invented for?  You would need to mock `clone()` to return a mock that expects a single call to `Setup()`?

Reply:
D'oh. I redid the test with mocks, and it no longer reads like a logic puzzle.

Also, I split it up into two separate tests: one for retry + backoff; one for polling.